File: python/queue/stack_use_queue.py

# 使用队列实现一个栈

class MyStack:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.input = []

    def push(self, x: int) -> None:
        """
        Push element x onto stack.
        """
        self.input.append(x)
        if len(self.input) > 0:
            self.input = [x] + self.input[:len(self.input) - 1]

    def pop(self) -> int:
        """
        Removes the element on top of the stack and returns that element.
        """
        return self.input.pop(0)

    def top(self) -> int:
        """
        Get the top element.
        """
        return self.input[0]


    def empty(self) -> bool:
        """
        Returns whether the stack is empty.
        """
        return len(self.input) == 0

File: java/datastructure/queue/leetcode/StackUseQueue.java
package datastructure.queue.leetcode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author roseduan
 * 使用队列实现栈
 */
public class StackUseQueue {

    private Queue<Integer> queue;
    
    /** Initialize your data structure here. */
    public StackUseQueue() {
        this.queue = new LinkedList<>();
    }
    
    /** Push element x onto stack. */
    public void push(int x) {
        int n = queue.size();
        queue.offer(x);
        for (int i = 0; i < n; i++) {
            queue.offer(queue.poll());
        }
    }
    
    /** Removes the element on top of the stack and returns that element. */
    public int pop() {
        return queue.poll();
    }
    
    /** Get the top element. */
    public int top() {
        return queue.peek();
    }
    
    /** Returns whether the stack is empty. */
    public boolean empty() {
        return queue.isEmpty();
    }
}
